<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подготовка" data-toc-modified-id="Подготовка-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подготовка</a></span><ul class="toc-item"><li><span><a href="#Установка-и-импорт-библиотек" data-toc-modified-id="Установка-и-импорт-библиотек-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Установка и импорт библиотек</a></span></li><li><span><a href="#Общая-информация-о-датасете" data-toc-modified-id="Общая-информация-о-датасете-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Общая информация о датасете</a></span></li><li><span><a href="#Баланс-классов" data-toc-modified-id="Баланс-классов-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Баланс классов</a></span></li><li><span><a href="#Очистка-текста-и-лемматизация" data-toc-modified-id="Очистка-текста-и-лемматизация-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Очистка текста и лемматизация</a></span></li><li><span><a href="#Подготовка-признаков-с-помощью-TF-IDF" data-toc-modified-id="Подготовка-признаков-с-помощью-TF-IDF-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Подготовка признаков с помощью TF-IDF</a></span><ul class="toc-item"><li><span><a href="#Разбиение-на-выборки" data-toc-modified-id="Разбиение-на-выборки-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Разбиение на выборки</a></span></li><li><span><a href="#Удаление-стоп-слов-и-получение-признаков" data-toc-modified-id="Удаление-стоп-слов-и-получение-признаков-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Удаление стоп-слов и получение признаков</a></span></li></ul></li><li><span><a href="#Подготовка-признаков-с-помощью-BERT" data-toc-modified-id="Подготовка-признаков-с-помощью-BERT-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Подготовка признаков с помощью BERT</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#TF-IDF" data-toc-modified-id="TF-IDF-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>TF-IDF</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.1.2"><span class="toc-item-num">2.1.2&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.1.3"><span class="toc-item-num">2.1.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#BERT" data-toc-modified-id="BERT-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>BERT</a></span><ul class="toc-item"><li><span><a href="#Логистическая-регрессия" data-toc-modified-id="Логистическая-регрессия-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Логистическая регрессия</a></span></li><li><span><a href="#Дерево-решений" data-toc-modified-id="Дерево-решений-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>Дерево решений</a></span></li><li><span><a href="#Случайный-лес" data-toc-modified-id="Случайный-лес-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>Случайный лес</a></span></li></ul></li><li><span><a href="#Лучшая-модель" data-toc-modified-id="Лучшая-модель-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Лучшая модель</a></span></li></ul></li><li><span><a href="#Выводы" data-toc-modified-id="Выводы-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы</a></span></li></ul></div>

# Выявление токсичных комментариев

## Подготовка

### Установка и импорт библиотек

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
!pip install transformers

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [3]:
import pandas as pd

import numpy as np

# очистка текста и лемматизация
import re 
import spacy

# удаление стоп-слов
from nltk.corpus import stopwords
import nltk

from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.model_selection import train_test_split
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import RandomizedSearchCV

from sklearn.utils import shuffle

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import f1_score

# BERT
import transformers
import torch
from tqdm import notebook

from sklearn.pipeline import make_pipeline, Pipeline

In [4]:
spacy.prefer_gpu()

True

In [5]:
torch.cuda.is_available()

True

In [6]:
device = torch.device("cuda:0") if torch.cuda.is_available() else torch.device("cpu")
device

device(type='cuda', index=0)

### Общая информация о датасете

In [7]:
full_data = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/toxic_comments.csv')
full_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


In [8]:
full_data.head()

text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0

**Вывод**

Датасет содержит 159571 строку - комментарии на английском языке. Стиль заголовков соблюден.

### Баланс классов

In [9]:
full_data['toxic'].value_counts(normalize=True)

0    0.898321
1    0.101679
Name: toxic, dtype: float64

In [10]:
def downsample(data, fraction):
    data_zeros = data.query('toxic==0')
    data_ones = data.query('toxic==1')

    data_sample = data_zeros.sample(frac=fraction, random_state=12345)
    
    data_downsampled = pd.concat([data_sample] + [data_ones])
    
    data_downsampled = shuffle(data_downsampled, random_state=12345)    
    
    return data_downsampled

In [11]:
data = downsample(full_data, 0.11).reset_index(drop=True)

In [12]:
data['toxic'].value_counts(normalize=True)

1    0.507142
0    0.492858
Name: toxic, dtype: float64

In [13]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31993 entries, 0 to 31992
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   text    31993 non-null  object
 1   toxic   31993 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 500.0+ KB


**Вывод**

В данных наблюдается дисбаланс классов. Т.к. выборка достаточно большая и лемматизация на всем этом объеме не выполняется (происходит зависание на имеющихся ресурсах), то применяю downsampling. Таким образом, получила выборку из 31993 строк вместо 159571.

### Очистка текста и лемматизация

In [14]:
lemmatizer = spacy.load('en_core_web_sm')

In [15]:
def clear_text(text_full):
    
    text = text_full.lower()
    text = re.sub(r"[^a-zA-Z]", ' ', text) # символы заменяются на пробелы
    text = ' '.join(text.split())
    
    doc = lemmatizer(text) # не включает пробелы
    text = ' '.join([token.lemma_ for token in doc])  # лемматизированная строка
        
    text = text.replace('-PRON-', '')
    text = ' '.join(text.split())

    return text

In [16]:
%%time
data['lemm_text'] = data['text'].apply(clear_text)

CPU times: user 9min 6s, sys: 2.66 s, total: 9min 9s
Wall time: 9min 16s


In [17]:
def clear_text_for_bert(text_full):
    
    text = re.sub(r"[^a-zA-Z]", ' ', text_full) # символы заменяются на пробелы
    text = ' '.join(text.split())

    return text

In [18]:
%%time
data['bert_text'] = data['text'].apply(clear_text_for_bert)

CPU times: user 675 ms, sys: 12 ms, total: 687 ms
Wall time: 686 ms


In [19]:
data.head(10)

text  toxic  \
0   ROZ LIPSCHITZ IS BLOCKED BECAUSE HE IS AN IDIOT!      1   
1  Yeah I believe it, how many times have I heard...      1   
2  What, you think this block will work better th...      1   
3  Phil Spector is a FUCKING MURDERER !!! You kno...      1   
4  What the f*** do you want at my talk page. I g...      1   
5  Welll well, who else would the Tatar turn to a...      0   
6  Camp Lockett Wiki page begun using Buffalo Sol...      0   
7  Conrad Devonshire \n\nHe's a proven sockpuppet...      1   
8  "\n\n Please reconsider \n\nAll I can say, rea...      0   
9  Proposed deletion of Disappearance of Megan St...      0   

                                           lemm_text  \
0      roz lipschitz be block because he be an idiot   
1  yeah I believe it how many time have I hear th...   
2  what you think this block will work well than ...   
3  phil spector be a fucking murderer you know it...   
4  what the f do you want at my talk page I get n...   
5  welll well who else would the tatar turn to ap...   
6  camp lockett wiki page begin use buffalo soldi...   
7  conrad devonshire he s a prove sockpuppetter I...   
8  please reconsider all I can say really be plea...   
9  propose deletion of disappearance of megan sta...   

                                           bert_text  
0    ROZ LIPSCHITZ IS BLOCKED BECAUSE HE IS AN IDIOT  
1  Yeah I believe it how many times have I heard ...  
2  What you think this block will work better tha...  
3  Phil Spector is a FUCKING MURDERER You know it...  
4  What the f do you want at my talk page I got n...  
5  Welll well who else would the Tatar turn to ap...  
6  Camp Lockett Wiki page begun using Buffalo Sol...  
7  Conrad Devonshire He s a proven sockpuppetter ...  
8  Please reconsider All I can say really is plea...  
9  Proposed deletion of Disappearance of Megan St...

**Вывод**

На этом этапе предобработки выполнила следующие манипуляции:
* преобразование слов к нижнему регистру;
* удаление всех символов, кроме латинских букв;
* лемматизация;
* удаление `-PRON-` (появляется вместо местоимений при лемматизации `spacy`).

Для BERT выполнила только удаление всех символов, кроме латинских букв, т.к. остальное модель выполняет сама.

### Подготовка признаков с помощью TF-IDF

#### Разбиение на выборки

In [20]:
corpus = data['lemm_text']

In [21]:
target = data['toxic']

In [22]:
corpus_train, corpus_test, tf_idf_target_train, tf_idf_target_test = train_test_split(corpus, target, test_size=0.2, random_state=12345)

In [23]:
corpus_train.shape, tf_idf_target_train.shape

((25594,), (25594,))

In [24]:
corpus_test.shape, tf_idf_target_test.shape

((6399,), (6399,))

**Вывод**

Выделила под тестовую выборку - 20% от датасета, полученного после уменьшения выборки.

#### Удаление стоп-слов и получение признаков

In [25]:
nltk.download('stopwords')
stop_words = set(stopwords.words('english'))
count_tf_idf = TfidfVectorizer(stop_words=stop_words) 

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


**Вывод**

Подготовлен векторизатор для получения матрицы TF-IDF без учета стоп-слов. Он будет добавлен в пайплайн для кросс-валидации.

### Подготовка признаков с помощью BERT

In [26]:
pretrained_weights = 'distilbert-base-uncased'
tokenizer = transformers.DistilBertTokenizer.from_pretrained(pretrained_weights)
model = transformers.DistilBertModel.from_pretrained(pretrained_weights)

Some weights of the model checkpoint at distilbert-base-uncased were not used when initializing DistilBertModel: ['vocab_layer_norm.weight', 'vocab_projector.weight', 'vocab_transform.weight', 'vocab_layer_norm.bias', 'vocab_projector.bias', 'vocab_transform.bias']
- This IS expected if you are initializing DistilBertModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing DistilBertModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).


In [27]:
tokenized = data['bert_text'].apply(lambda x: tokenizer.encode(x, add_special_tokens=True, truncation=True))

n = max(tokenized.apply(lambda x: len(x)))
# англ. вектор с отступами
padded = np.array([i + [0]*(n - len(i)) for i in tokenized.values])

attention_mask = np.where(padded != 0, 1, 0)

In [28]:
padded.shape

(31993, 512)

In [29]:
attention_mask.shape

(31993, 512)

In [30]:
batch_size = 100

embeddings = []

# добавила +1, чтобы брались записи, ктр не попадали из-за некратности общего количества записей числу 100
for i in notebook.tqdm(range(padded.shape[0] // batch_size + 1)): 
    # преобразуем данные
    batch = torch.LongTensor(padded[batch_size*i:batch_size*(i+1)]).to(device)
    # преобразуем маску
    attention_mask_batch = torch.LongTensor(attention_mask[batch_size*i:batch_size*(i+1)]).to(device)

    with torch.no_grad():
        model.to(device)
        batch_embeddings = model(batch, attention_mask=attention_mask_batch)

    # Из полученного тензора извлечём нужные элементы и добавим в список всех эмбеддингов
    # преобразуем элементы методом numpy() к типу numpy.array
    embeddings.append(batch_embeddings[0][:,0,:].cpu().numpy())

    del batch
    del attention_mask_batch
    del batch_embeddings

  0%|          | 0/320 [00:00<?, ?it/s]

In [31]:
# Соберём все эмбеддинги в матрицу признаков вызовов функции concatenate()
bert_features = np.concatenate(embeddings)
bert_features.shape

(31993, 768)

In [32]:
bert_target = data['toxic']
bert_target.shape

(31993,)

In [33]:
bert_features_train, bert_features_test, bert_target_train, bert_target_test = train_test_split(bert_features, bert_target, test_size=0.2, random_state=12345)

In [34]:
bert_features_train.shape, bert_target_train.shape

((25594, 768), (25594,))

In [35]:
bert_features_test.shape, bert_target_test.shape

((6399, 768), (6399,))

**Вывод**

Для готового очищенного и лемматизированного текста применила модель `DistilBertModel` для получения признаков.  

Также выделила тестовую выборку в 20%.

## Обучение

In [36]:
metrics = pd.DataFrame(columns = ['nlp_type', 'model_type', 'depth', 'n_estimators', 'f1_score', 'best_model'])

In [37]:
MIN_DEPTH = 10
MAX_DEPTH = 20
MAX_DEPTH_TREE = 30
CV = 4 # тестовая выборка 20%, поэтому валдационная тоже при кросс-валидации пусть будет 20% датасета
MIN_EST = 1
MAX_EST = 50
N_ITER = 4    

DISTRIBUTIONS_FOREST = {'max_depth':range(MIN_DEPTH, MAX_DEPTH + 1, 5), 'n_estimators':range(MIN_EST,MAX_EST + 1, 10)}
DISTRIBUTIONS_TREE = {'max_depth':range(1, MAX_DEPTH_TREE + 1, 5)} 

In [38]:
def get_best_params(model, features_train, target_train, distributions=None, verbose=None):
    
    if distributions==None:
        distributions = DISTRIBUTIONS_FOREST
      
    grid = RandomizedSearchCV(model, distributions, cv = CV, scoring = 'f1', n_iter=N_ITER, 
                              n_jobs = -1, random_state=12345)
    
    if verbose==None:
        grid.fit(features_train, target_train)
    else:
        grid.fit(features_train, target_train, verbose=verbose)
    
    return grid.best_estimator_, round(grid.best_score_, 3)

### TF-IDF

#### Логистическая регрессия

In [39]:
%%time
model = LogisticRegression(random_state=12345, solver='liblinear')
tf_idf_pipeline = make_pipeline(count_tf_idf, model)

scores = cross_val_score(tf_idf_pipeline, corpus_train, tf_idf_target_train, scoring='f1', cv=CV)
metrics.loc[len(metrics.index)] = ['TF-IDF', 'LogisticRegression', None, None, round(pd.Series(scores).mean(), 3), tf_idf_pipeline]

CPU times: user 4.37 s, sys: 833 ms, total: 5.21 s
Wall time: 4.33 s


#### Дерево решений

In [40]:
%%time
model = DecisionTreeClassifier(random_state=12345)
tf_idf_pipeline = Pipeline([
      ('count_tf_idf', count_tf_idf), 
      ('tree',model)
  ])

pipeline_distributions = {'tree__' + key: DISTRIBUTIONS_TREE[key] for key in DISTRIBUTIONS_TREE}

best_model, best_score = get_best_params(tf_idf_pipeline, corpus_train, tf_idf_target_train, pipeline_distributions)
metrics.loc[len(metrics.index)] = ['TF-IDF', 'DecisionTreeClassifier', best_model['tree'].max_depth, None, best_score, best_model] 

CPU times: user 3.93 s, sys: 298 ms, total: 4.23 s
Wall time: 30.4 s


#### Случайный лес

In [41]:
%%time

model = RandomForestClassifier(random_state=12345)
tf_idf_pipeline = Pipeline([
      ('count_tf_idf', count_tf_idf), 
      ('forest',model)
  ])

pipeline_distributions = {'forest__' + key: DISTRIBUTIONS_FOREST[key] for key in DISTRIBUTIONS_FOREST}

best_model, best_score = get_best_params(tf_idf_pipeline, corpus_train, tf_idf_target_train, pipeline_distributions)
metrics.loc[len(metrics.index)] = ['TF-IDF', 'RandomForestClassifier', best_model['forest'].max_depth, 
                                   best_model['forest'].n_estimators, best_score, best_model] 

CPU times: user 1.89 s, sys: 80.2 ms, total: 1.97 s
Wall time: 16.4 s


### BERT

#### Логистическая регрессия

In [42]:
%%time
model = LogisticRegression(random_state=12345, solver='liblinear')
scores = cross_val_score(model, bert_features_train, bert_target_train, scoring='f1', cv=CV)
metrics.loc[len(metrics.index)] = ['BERT', 'LogisticRegression', None, None, round(pd.Series(scores).mean(), 3), model]

CPU times: user 36.5 s, sys: 482 ms, total: 37 s
Wall time: 36.4 s


#### Дерево решений

In [43]:
%%time
model = DecisionTreeClassifier(random_state=12345)
best_model, best_score = get_best_params(model, bert_features_train, bert_target_train, DISTRIBUTIONS_TREE)
metrics.loc[len(metrics.index)] = ['BERT', 'DecisionTreeClassifier', best_model.max_depth, None, best_score, best_model] 

CPU times: user 34.5 s, sys: 282 ms, total: 34.7 s
Wall time: 4min 15s


#### Случайный лес

In [44]:
%%time
model = RandomForestClassifier(random_state=12345)
best_model, best_score = get_best_params(model, bert_features_train, bert_target_train)
metrics.loc[len(metrics.index)] = ['BERT', 'RandomForestClassifier', best_model.max_depth, best_model.n_estimators, best_score, 
                                   best_model] 

CPU times: user 20.6 s, sys: 126 ms, total: 20.7 s
Wall time: 1min 42s


### Лучшая модель

In [45]:
metrics = metrics.sort_values('f1_score', ascending=False).reset_index(drop=True)
best_model = metrics.loc[0, 'best_model']
metrics = metrics.drop('best_model', axis=1)
metrics

nlp_type              model_type depth n_estimators  f1_score
0     BERT      LogisticRegression  None         None     0.899
1   TF-IDF      LogisticRegression  None         None     0.889
2     BERT  RandomForestClassifier    15           31     0.867
3     BERT  DecisionTreeClassifier    16         None     0.797
4   TF-IDF  RandomForestClassifier    15           31     0.788
5   TF-IDF  DecisionTreeClassifier    26         None     0.744

In [46]:
best_model

LogisticRegression(random_state=12345, solver='liblinear')

**Вывод**

Обучила 6 моделей - 3 для TF-IDF, 3 для BERT - логистическая регрессия, дерево решений, случайный лес. Модели градиентного бустинга не использовала из-за большого времени выполнения.

В итоге лучшая F1-мера при кросс-валидации на модели логистической регрессии для признаков, подготовленных с помощью BERT.

## Выводы

In [47]:
if metrics.loc[0, 'nlp_type'] == 'BERT':
  best_model.fit(bert_features_train, bert_target_train)
  predictions = best_model.predict(bert_features_test)
  print(f1_score(bert_target_test, predictions))
else:
  best_model.fit(corpus_train, tf_idf_target_train)
  predictions = best_model.predict(corpus_test)
  print(f1_score(tf_idf_target_test, predictions))

0.8986645718774549


**Вывод**

Лучшая модель - логистическая регрессия для признаков, подготовленных с помощью BERT. Она дает на тестовой выборке F1-меру, равную 0.899.

При этом предобработка перед обучением (подготовка признаков для BERT и очистка текста с лемматизацией для TF_IDF) занимает соизмеримое время - в пределах 10 минут для выборки после downsampling, обучение моделей с кросс-валидацией в случае BERT заметно дольше (для случайного леса - 16 секунд против 104), а качество на кросс-валидации отличается на сотые.